openlibrary/solr/query_utils.py

```diff
@@ -0,0 +1,206 @@
+from typing import Callable, Optional
+from luqum.parser import parser
+from luqum.tree import Item, SearchField, BaseOperation, Group, Word, Unary
+import re
+
+
+class EmptyTreeError(Exception):
+    pass
+
+
+def luqum_remove_child(child: Item, parents: list[Item]):
+    """
+    Removes a child from a luqum parse tree. If the tree
+    ends up being empty, errors.
+
+    :param child: Node to remove
+    :param parents: Path of parent nodes leading from the root of the tree
+    """
+    parent = parents[-1] if parents else None
+    if parent is None:
+        # We cannot remove the element if it is the root of the tree
+        raise EmptyTreeError()
+    elif isinstance(parent, (BaseOperation, Group, Unary)):
+        new_children = tuple(c for c in parent.children if c != child)
+        if not new_children:
+            # If we have deleted all the children, we need to delete the parent
+            # as well. And potentially recurse up the tree.
+            luqum_remove_child(parent, parents[:-1])
+        else:
+            parent.children = new_children
+    else:
+        raise ValueError("Not supported for generic class Item")
+
+
+def luqum_traverse(item: Item, _parents: list[Item] = None):
+    """
+    Traverses every node in the parse tree in depth-first order.
+
+    Does not make any guarantees about what will happen if you
+    modify the tree while traversing it 😅 But we do it anyways.
+
+    :param item: Node to traverse
+    :param _parents: Internal parameter for tracking parents
+    """
+    parents = _parents or []
+    yield item, parents
+    new_parents = [*parents, item]
+    for child in item.children:
+        yield from luqum_traverse(child, new_parents)
```

Comment:
If we did hit some interesting performance characteristic in complex queries, this may be a place to investigate?

Reply:
Unlikely? Unless we have an infinite recur somewhere